File: src/main/java/hiiragi283/material/item/ItemBookRespawn.java
package hiiragi283.material.item;

import hiiragi283.material.api.item.HiiragiItem;
import hiiragi283.material.util.HiiragiUtil;
import net.minecraft.entity.player.EntityPlayer;
import net.minecraft.item.EnumRarity;
import net.minecraft.item.ItemStack;
import net.minecraft.util.ActionResult;
import net.minecraft.util.EnumHand;
import net.minecraft.world.World;
import net.minecraftforge.common.IRarity;
import org.jetbrains.annotations.NotNull;

import javax.annotation.ParametersAreNonnullByDefault;

public class ItemBookRespawn extends HiiragiItem {

    public ItemBookRespawn() {
        super("book_respawn", 0);
    }

    //    General    //

    @Override
    public @NotNull IRarity getForgeRarity(@NotNull ItemStack stack) {
        return EnumRarity.EPIC;
    }

    //    Event    //

    @Override
    @ParametersAreNonnullByDefault
    public @NotNull ActionResult<ItemStack> onItemRightClick(World world, EntityPlayer player, EnumHand hand) {
        //落下死防止やコマンド権限のためクリエモードに切り替え
        if (!world.isRemote) HiiragiUtil.executeCommand(player, "gamemode 1");
        if (world.isRemote) {
            var spawnPoint = world.getSpawnPoint();
            player.motionX = 0.0;
            player.motionY = 0.0;
            player.motionZ = 0.0; //運動ベクトルをリセット
            //プレイヤーを指定した座標にテレポート
            player.setLocationAndAngles(spawnPoint.getX() + 0.5, 128.0, spawnPoint.getZ() + 0.5, 0.0f, 0.0f);
        }
        return super.onItemRightClick(world, player, hand);
    }

}

File: src/main/kotlin/hiiragi283/material/item/ItemBookRespawn.kt
package hiiragi283.material.item

import hiiragi283.material.api.item.HiiragiItem
import hiiragi283.material.api.module.IRecipeModuleItem
import hiiragi283.material.api.recipe.IMachineRecipe
import hiiragi283.material.util.executeCommand
import net.minecraft.entity.player.EntityPlayer
import net.minecraft.item.EnumRarity
import net.minecraft.item.ItemStack
import net.minecraft.util.ActionResult
import net.minecraft.util.EnumHand
import net.minecraft.world.World
import net.minecraftforge.common.IRarity

object ItemBookRespawn : HiiragiItem("book_respawn", 0), IRecipeModuleItem {

    //    General    //

    override fun getForgeRarity(stack: ItemStack): IRarity = EnumRarity.EPIC

    //    Event    //

    override fun onItemRightClick(world: World, player: EntityPlayer, hand: EnumHand): ActionResult<ItemStack> {
        //落下死防止やコマンド権限のためクリエモードに切り替え
        if (!world.isRemote) executeCommand(player, "gamemode 1")
        if (world.isRemote) {
            val spawnPoint = world.spawnPoint
            player.motionX = 0.0
            player.motionY = 0.0
            player.motionZ = 0.0 //運動ベクトルをリセット
            //プレイヤーを指定した座標にテレポート
            player.setLocationAndAngles(spawnPoint.x + 0.5, 128.0, spawnPoint.z + 0.5, 0.0f, 0.0f)
        }
        return super.onItemRightClick(world, player, hand)
    }

    //    IRecipeModuleItem    //

    override val recipeType: IMachineRecipe.Type = IMachineRecipe.Type.TEST

}